components/snap_recovery/src/services.rs

```diff
@@ -0,0 +1,350 @@
+    for cf_name in engine.cf_names() {
+        let cf = cf_name.to_owned().clone();
+        let kv_db = engine.clone();
+        let h = Builder::new()
+            .name(format!("compact-{}", cf))
+            .spawn_wrapper(move || {
+                info!("recovery starts manual compact"; "cf" => cf.clone());
+                tikv_alloc::add_thread_memory_accessor();
+                let db = kv_db.as_inner();
+                let handle = get_cf_handle(db, cf.as_str()).unwrap();
+                let mut compact_opts = CompactOptions::new();
+                compact_opts.set_max_subcompactions(64);
+                compact_opts.set_exclusive_manual_compaction(false);
+                compact_opts.set_bottommost_level_compaction(DBBottommostLevelCompaction::Skip);
+                db.compact_range_cf_opt(handle, &compact_opts, None, None);
+                tikv_alloc::remove_thread_memory_accessor();
+
+                info!("recovery finishes manual compact"; "cf" => cf);
+            })
+            .expect("failed to spawn compaction thread");
+        handles.push(h);
+    }
+    for h in handles {
+        h.join()
+            .unwrap_or_else(|e| error!("thread handle join error"; "error" => ?e));
+    }
+    Ok(())
+}
+
+impl<ER: RaftEngine> RecoverData for RecoveryService<ER> {
+    // 1. br start to ready region meta
+    fn read_region_meta(
+        &mut self,
+        _ctx: RpcContext<'_>,
+        _req: ReadRegionMetaRequest,
+        mut sink: ServerStreamingSink<RegionMeta>,
+    ) {
+        let (tx, rx) = mpsc::unbounded();
+        // tx only clone once within RegionMetaCollector, so that it drop automatically
+        // when work thread done
+        let meta_collector = RegionMetaCollector::new(self.engines.clone(), tx);
+        info!("start to collect region meta");
+        meta_collector.start_report();
+        let send_task = async move {
+            let mut s = rx.map(|resp| Ok((resp, WriteFlags::default())));
+            sink.send_all(&mut s).await?;
+            sink.close().await?;
+            Ok(())
+        }
+        .map(|res: std::result::Result<(), gRPCError>| match res {
+            Ok(_) => {
+                info!("collect region meta done");
+            }
+            Err(e) => {
+                error!("rcollect region meta failure"; "error" => ?e);
+            }
+        });
+
+        self.threads.spawn_ok(send_task);
+    }
+
+    // 2. br start to recover region
+    // assign region leader and wait leader apply to last log
+    fn recover_region(
+        &mut self,
+        _ctx: RpcContext<'_>,
+        mut stream: RequestStream<RecoverRegionRequest>,
+        sink: ClientStreamingSink<RecoverRegionResponse>,
+    ) {
+        let raft_router = self.router.clone();
+        let store_id = self.get_store_id();
+        info!("start to recover the region");
+        let task = async move {
+            let mut leaders = Vec::new();
+            while let Some(req) = stream.next().await {
```

Comment:
It's better to make all logic async executed in the loop. Maybe we can refactor this function later.

Reply:
Ok

Reply:
will make a TODO refactor this part.